From fb7fe190e1d43961f94ba4dc4c39254b1b6b1be0 Mon Sep 17 00:00:00 2001
From: Pankaj Maharana <pankaj.maharana@lge.com>
Date: Thu, 12 Dec 2019 12:58:59 +0530
Subject: [PATCH] v4l2videodec: Check stop in flush() to avoid race

 condition

Backward playback will drain and flush every frame. Stop playback
when backward playback have race condition between exit thread and
streaming thread flush. Add one check to avoid it.

Fixes #639

Upstream-Status: Submitted
https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/commit/9d42b2dd5dea81d0b8f46cb363be954aaaa34d4c

---
 sys/v4l2/gstv4l2videodec.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index 8c06c90..afebf56 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -347,6 +347,9 @@ gst_v4l2_video_dec_flush (GstVideoDecoder * decoder)
 
   self->output_flow = GST_FLOW_OK;
 
+  if (G_UNLIKELY (!g_atomic_int_get (&self->active)))
+    return TRUE;
+
   gst_v4l2_object_unlock_stop (self->v4l2output);
   gst_v4l2_object_unlock_stop (self->v4l2capture);
 
